File: past-competitions/2018/src/brett-python/Prob19.py
# Define the rotors
# Each value represents the amount that must be added
# to get the output value when feeding input into that index.
# Index values must be adjusted to account for rotation.
rotors = [[1, 2, 4, 7, 1, 9, 2, 0, 1, 3],
          [0, 2, 3, 9, 2, 4, 5, 7, 0, 8],
          [5, 8, 9, 4, 9, 3, 4, 5, 6, 7],
          [1, 5, 3, 9, 5, 5, 1, 7, 5, 9]]
inverseRotors = [[3, 9, 7, 8, 1, 9, 6, 0, 8, 9],
                 [0, 5, 1, 8, 3, 7, 8, 2, 0, 6],
                 [6, 1, 5, 1, 4, 5, 3, 6, 7, 2],
                 [5, 9, 1, 5, 3, 7, 5, 9, 1, 5]]
# Reflector doesn't rotate
reflector = [3, 5, 6, 7, 1, 9, 5, 2, 4, 8]

# Open the input file
with open("Prob19.in.txt", "rt") as inputFile:
    # Read the number of test cases (trim out the newline)
    cases = int(inputFile.readline().replace("\n", ""))
    # For each test case
    for caseNum in range(cases):
        # Read the key settings
        leftData = inputFile.readline().replace("\n", "")
        middleData = inputFile.readline().replace("\n", "")
        rightData = inputFile.readline().replace("\n", "")
        leftRotorData = leftData.split(" ")
        middleRotorData = middleData.split(" ")
        rightRotorData = rightData.split(" ")

        # Convert to integers
        leftRotor = int(leftRotorData[0]) - 1
        middleRotor = int(middleRotorData[0]) - 1
        rightRotor = int(rightRotorData[0]) - 1
        leftRotorStart = int(leftRotorData[1])
        middleRotorStart = int(middleRotorData[1])
        rightRotorStart = int(rightRotorData[1])
        
        # Read plaintext
        plaintext = inputFile.readline().replace("\n", "")

        # Set up cipher
        leftPosition = leftRotorStart
        middlePosition = middleRotorStart
        rightPosition = rightRotorStart
        for i in range(len(plaintext)):
            value = int(plaintext[i])

            # Enter left rotor
            # Calculate index for left rotor (same process for each rotor)
            index = value - leftPosition
            if index < 0:
                index = index + 10
            # Add the value of the calculated index.
            # Mod by 10 to keep single digit
            value = (value + rotors[leftRotor][index]) % 10

            # Enter middle rotor
            index = value - middlePosition
            if index < 0:
                index = index + 10
            value = (value + rotors[middleRotor][index]) % 10

            # Enter right rotor
            index = value - rightPosition
            if index < 0:
                index = index + 10
            value = (value + rotors[rightRotor][index]) % 10

            # Enter reflector
            # This doesn't rotate, so value is the index
            value = (value + reflector[value]) % 10

            # Now go backward through the rotors
            # Reenter right rotor
            index = value - rightPosition
            if index < 0:
                index = index + 10
            value = (value + inverseRotors[rightRotor][index]) % 10

            # Reenter middle rotor
            index = value - middlePosition
            if index < 0:
                index = index + 10
            value = (value + inverseRotors[middleRotor][index]) % 10

            # Reenter left rotor
            index = value - leftPosition
            if index < 0:
                index = index + 10
            value = (value + inverseRotors[leftRotor][index]) % 10

            # Print the resulting digit, suppressing the newline
            print(str(value), end="")

            # Rotate the right rotor one position
            # When it returns to its original position, rotate
            # the middle rotor one position, and so on
            rightPosition = (rightPosition + 1) % 10
            if rightPosition == rightRotorStart:
                middlePosition = (middlePosition + 1) % 10
                if middlePosition == middleRotorStart:
                    leftPosition = (leftPosition + 1) % 10
            # end rotation if
        # end for loop
        # Print newline for next test case
        print("")

File: past-competitions/2018/src/brett-cpp/Prob19.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char *argv[])
{
	// open file for reading
	FILE *inputFile = fopen("Prob19.in.txt", "r");
	int testCases;

	// read number of test cases
	fscanf(inputFile, "%d\n", &testCases);
	
	char nextChar;
	int currentValue;

	// each number represents the amount to add to the input
	// to get the output. each rotor is listed in position 0.
	// index values must be adjusted to account for rotation.
	int rotors[4][10] = {{1, 2, 4, 7, 1, 9, 2, 0, 1, 3},
			{0, 2, 3, 9, 2, 4, 5, 7, 0, 8},
			{5, 8, 9, 4, 9, 3, 4, 5, 6, 7},
			{1, 5, 3, 9, 5, 5, 1, 7, 5, 9}};
	int inverseRotors[4][10] = {{3, 9, 7, 8, 1, 9, 6, 0, 8, 9},
			{0, 5, 1, 8, 3, 7, 8, 2, 0, 6},
			{6, 1, 5, 1, 4, 5, 3, 6, 7, 2},
			{5, 9, 1, 5, 3, 7, 5, 9, 1, 5}};
	int reflector[10] = {3, 5, 6, 7, 1, 9, 5, 2, 4, 8};

	int leftRotor, leftRotorStart, leftRotorPosition;
	int middleRotor, middleRotorStart, middleRotorPosition;
	int rightRotor, rightRotorStart, rightRotorPosition;
	int rotorIndex;

	// execute test cases
	while(testCases > 0){
		// read key settings
		fscanf(inputFile, "%d %d\n", &leftRotor, &leftRotorStart);
		fscanf(inputFile, "%d %d\n", &middleRotor, &middleRotorStart);
		fscanf(inputFile, "%d %d\n", &rightRotor, &rightRotorStart);

		// adjust for index
		leftRotor = leftRotor - 1;
		middleRotor = middleRotor - 1;
		rightRotor = rightRotor - 1;
		// set start positions
		leftRotorPosition = leftRotorStart;
		middleRotorPosition = middleRotorStart;
		rightRotorPosition = rightRotorStart;

		// read characters one at a time
		fscanf(inputFile, "%c", &nextChar);
		// until we hit a newline
		while(nextChar != '\n'){
			// convert from number - '0' == 48, '1' == 49
			// so '0' - 48 = 0, etc.
			currentValue = nextChar - 48;

			// enter left rotor
			// when in position 0, currentValue is the rotorIndex
			// when the rotor is shifted, we have to adjust accordingly
			rotorIndex = currentValue - leftRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			// add the rotor's value to the current value
			currentValue = currentValue + rotors[leftRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// enter middle rotor
			rotorIndex = currentValue - middleRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			currentValue = currentValue + rotors[middleRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// enter right rotor
			rotorIndex = currentValue - rightRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			currentValue = currentValue + rotors[rightRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// enter reflector
			// the reflector doesn't rotate
			currentValue = currentValue + reflector[currentValue];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// re-enter right rotor
			// same as above, but now we use the inverseRotor arrays
			// for going "backwards" through the rotors
			rotorIndex = currentValue - rightRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			currentValue = currentValue + inverseRotors[rightRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// re-enter middle rotor
			rotorIndex = currentValue - middleRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			currentValue = currentValue + inverseRotors[middleRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// re-enter left rotor
			rotorIndex = currentValue - leftRotorPosition;
			if(rotorIndex < 0){
				rotorIndex = rotorIndex + 10;
			}
			currentValue = currentValue + inverseRotors[leftRotor][rotorIndex];
			if(currentValue >= 10){
				currentValue = currentValue - 10;
			}

			// print the resulting value
			printf("%d", currentValue);

			// rotate rotors
			rightRotorPosition = rightRotorPosition + 1;
			if(rightRotorPosition == 10){
				rightRotorPosition = 0;
			}
			// if right rotor has completed a full rotation,
			if(rightRotorPosition == rightRotorStart){
				// rotate the middle one
				middleRotorPosition = middleRotorPosition + 1;
				if(middleRotorPosition == 10){
					middleRotorPosition = 0;
				}
				// if it's completed a full rotation,
				if(middleRotorPosition == middleRotorStart){
					// rotate the left one too
					leftRotorPosition = leftRotorPosition + 1;
					if(leftRotorPosition == 10){
						leftRotorPosition = 0;
					}
				}
			}

			// get next character
			fscanf(inputFile, "%c", &nextChar);
		}
		printf("\n");

		testCases = testCases - 1;
	}

	// close the file
	fclose(inputFile);
}
